activerecord/lib/active_record/dynamic_matchers.rb

```diff
@@ -61,11 +62,13 @@ def valid?
         end
 
         def define
-          model.class_eval <<-CODE, __FILE__, __LINE__ + 1
-            def self.#{name}(#{signature})
+          anonymous_module_with_method = Module.new
+          anonymous_module_with_method.class_eval <<-CODE, __FILE__, __LINE__ + 1
+            def #{name}(#{signature})
               #{body}
             end
           CODE
+          model.extend(anonymous_module_with_method)
```

Comment:
This will blow up the method cache since it is going to extend a module at runtime. I think we should implement this like the generated association methods, define the methods at runtime, but extend the module at load time.

Reply:
@rafaelfranca I changed the implementation to extend the module at load time and define the methods on the module at runtime, so method caching should work now, but the trouble is, since the method gets to the class through inheritance, `remove_method` doesn't work on it, so a bunch of tests fails. Do you have any tips for a workaround?

Reply:
Have you tried to implement it with something like https://github.com/rails/rails/blob/e6ef1fe056d7da084a4e8d3b817223e9a28e6200/activerecord/lib/active_record/core.rb#L223? That way you can access the module and remove the method from it.

Reply:
I did it somewhat differently, but this should do the trick indeed. Thanks for the advice!